common/src/api/internal/shared.rs

```diff
@@ -704,24 +704,48 @@ pub struct ResolvedVpcRouteSet {
 
 /// Describes the purpose of the dataset.
 #[derive(
-    Debug, Serialize, Deserialize, JsonSchema, Clone, Copy, PartialEq, Eq,
+    Debug,
+    Serialize,
+    Deserialize,
+    JsonSchema,
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    Ord,
+    PartialOrd,
+    Hash,
 )]
-#[serde(rename_all = "snake_case")]
+#[serde(tag = "type", rename_all = "snake_case")]
 pub enum DatasetKind {
-    Crucible,
+    #[serde(rename = "cockroach_db")]
     Cockroach,
+    Crucible,
     Clickhouse,
     ClickhouseKeeper,
     ExternalDns,
     InternalDns,
 }
 
+impl DatasetKind {
```

Comment:
A chunk of this implementation was moved from `sled-storage/src/dataset.rs`.

History lesson: We used to have types defined that would be usable by Nexus (`common/src/api/internal`), and types that were internal to the sled agent (`sled-storage/src/dataset.rs`) for defining dataset types/kinds.

This PR merges both concepts, but tweaks some names to avoid changing any schemas that are saved on M.2s.